London Releases

holgerd77 opened the pull request:
This PR prepares `london` compatible releases on the following libraries:

- `@ethereumjs/common`
- `@ethereumjs/tx`
- `@ethereumjs/block`
- `@ethereumjs/blockchain`
- `@ethereumjs/vm`

holgerd77 commented:
Ok, I am done with a first round on the `london` release PR. 😄 
(of course this can be updated on necessary final tweaks and fixes)

This PR can already be reviewed though to decouple and slowly get ready, think the vast majority of this should be stable. 🦀  🐠  🐡

holgerd77 commented on packages/block/README.md:
```diff
@@ -53,15 +55,31 @@ try {
+
+EIP-1559 blocks have an extra `baseFeePerGas` field (default: `new BN(7)`) and can encompass `FeeMarketEIP1559Transaction` txs (type `2`) (supported by `@ethereumjs/tx` `v3.2.0` or higher) as well as  `Transaction` legacy txs (internal type `0`) and `AccessListEIP2930Transaction` txs (type `1`).
+
+## Consensus Types
```
This diff here looks a bit strange, better to be viewed as a complete file in rendered format.

ryanio commented on packages/tx/README.md:
```diff
@@ -33,17 +33,52 @@ All types of transaction objects are frozen with `Object.freeze()` which gives y
+- Activation: `london`
+- Type: `2`
+
+This is the recommended tx type starting with the activation of the `london` HF, see the following code snipped for an example on how to instantiate:
```
```suggestion
This is the recommended tx type starting with the activation of the `london` HF, see the following code snippet for an example on how to instantiate:
```

ryanio commented on packages/tx/README.md:
```diff
@@ -33,17 +33,52 @@ All types of transaction objects are frozen with `Object.freeze()` which gives y
+- Type: `1`
 
-This is the recommended tx type starting with the activation of the `berlin` HF, see the following code snipped for an example on how to instantiate:
+This transaction type has been introduced along the `berlin` HF. See the following code snipped for an example on how to instantiate:
```
```suggestion
This transaction type has been introduced along the `berlin` HF. See the following code snippet for an example on how to instantiate:
```

ryanio commented on packages/vm/CHANGELOG.md:
```diff
@@ -6,6 +6,68 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+
+#### EIP-1559: Gas Fee Market
+
+The VM can now run `EIP-1559` compatible blocks (introduced with the `@ethereumjs/block` `v3.3.0` release) with `VM.runBlocks()` as well as `EIP-1559` txs with type `2` (introduced along the `@ethereumjs/tx` `v3.2.0` release), which can now be passed to `VM.runTx()` as the tx to be executed. Block and tx validation is happening accordingly and the gas calculation takes the new gas fee market para ...
```
```suggestion
The VM can now run `EIP-1559` compatible blocks (introduced with the `@ethereumjs/block` `v3.3.0` release) with `VM.runBlock()` as well as `EIP-1559` txs with type `2` (introduced along the `@ethereumjs/tx` `v3.2.0` release), which can now be passed to `VM.runTx()` as the tx to be executed. Block and tx validation is happening accordingly and the gas calculation takes the new gas fee market parameters from the block (`baseFeePerGas`) and the tx(s) (`maxFeePerGas` and `maxPriorityFeePerGas` instead of a `gasPrice`) into account.
```

ryanio commented on packages/vm/CHANGELOG.md:
```diff
@@ -6,6 +6,68 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+
+### StateManager: Preserve State History
+
+This VM release bumps the `merkle-patricia-tree` dependeny to `v4.2.0`, which is used as a datastore for the default `StateManager` implementation. The new MPT version switches to a default behavior to not delete any trie nodes on checkpoint commits, which has implications on the `StateManager.commit()` function which internally calls the MPT commit. This allows to go back to older trie states by ...
```
```suggestion
This VM release bumps the `merkle-patricia-tree` dependency to `v4.2.0`, which is used as a datastore for the default `StateManager` implementation. The new MPT version switches to a default behavior to not delete any trie nodes on checkpoint commits, which has implications on the `StateManager.commit()` function which internally calls the MPT commit. This allows to go back to older trie states by setting a new (old) state root with `StateManager.setStateRoot()`. The trie state is now guaranteed to still be consistent and complete, which has not been the case before and lead to erroneous behavior in certain usage scenarios (e.g. reported by HardHat).
```

ryanio commented on packages/vm/CHANGELOG.md:
```diff
@@ -6,6 +6,68 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+
+### Error Handling: Correct Non-VM Error Propagation
+
+In former versions of the VM non-VM errors happing inside the VM have been (unintentionally) shielded by a `try / catch` clause in the VM `Interpreter` class. This lead to existing bugs being hidden and channeled through as VM errors, which made it extremely difficult to trace such bugs down to the root cause. These kind of errors are now properly propagated and therefore lead to a break of the V ...
```
```suggestion
In former versions of the VM non-VM errors happening inside the VM have been (unintentionally) shielded by a `try / catch` clause in the VM `Interpreter` class. This lead to existing bugs being hidden and channeled through as VM errors, which made it extremely difficult to trace such bugs down to the root cause. These kind of errors are now properly propagated and therefore lead to a break of the VM control flow. Please note that this might lead to your code breaking *if* you have got an error in your implementation (this should be a good this though since now this bug can finally be fixed 😀 ).
```

holgerd77 commented:
@ryanio thanks for the updates here! 😄 

Rebased this and added a note on #1264

holgerd77 commented:
Ok, I have addressed all review requests from @jochem-brouwer on my additional changes on #1048, hopefully sufficient, for some I have opened somewhat extensive new issues.

Have also done a last `baikal` sync (up to block 37.000) without issues. I would therefore now put this in final review. 😄 

I would pledge for an early release here, tomorrow (Wednesday) or even still today (not sure if realistic). If we still release early in the week ecosystem devs would be able to eventually already start experimenting and integrating *this* week, this wouldn't be the case any more if we release on Thursday or Friday. And I think this would be an important week to gain at this stage.

We could then continue at some relaxed and solid pace and eventually do another round of releases in two weeks or so with some eventually necessary last fixes or tweaks. Then the final `ethereum/tests` releases should also be out and it should be possible to run the libraries against the final test suite.

I we start now with running some last minute tests against some open 1559 branches from there I think the chance (risk) is high that we get caught up into some hastily fix discussions where failures might even be triggered by tests not being finalized.

And for our mainly dev-centric user audience I would judge this more important to get people have their hands on testing the base functionality and start integrating than us double-ensuring that we have fixed the last very edgy bug.

Ok, if you go along with this arguments I would be happy on some approval here. 🙂 

If not drop a note. 😋

holgerd77 commented:
Update: if you have last change requests to the release PR itself please directly update the PR! Thanks!

ryanio reviewed (approved):
LGTM!

holgerd77 commented:
Just published the following releases on npm:

- `@ethereumjs/common ` [v2.3.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fcommon%402.3.0)
- `@ethereumjs/tx` [v3.2.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Ftx%403.2.0)
- `@ethereumjs/block` [v3.3.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fblock%403.3.0)
- `@ethereumjs/blockchain` [v5.3.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fblockchain%405.3.0)
- `@ethereumjs/vm` [v5.4.0](https://github.com/ethereumjs/ethereumjs-monorepo/releases/tag/%40ethereumjs%2Fvm%405.4.0)
